Update documentation using deterministic generation

rustyrussell opened the pull request:
~Depends on #8556~ Merged!

Now we do major surgery on autogenerate-rpc-examples.py.  Finally we run it to produce the canned blocks and wallet for future runs, and re-enable it in CI.

rustyrussell committed:
lightningd: cancel watching original funding when we switch to the new one via splice.

This happens if the channel is *not* announcable yet.  Then we hit the assertion
in funding_depth_cb that the txid is the same as the current funding.txid.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
Changelog-EXPERIMENTAL: fixed crash when we splice a channel which hasn't been announced yet.

rustyrussell committed:
common: add randbytes() wrapper to override cryptographic entropy: $CLN_DEV_ENTROPY_SEED

Only in developer mode, ofc.

Notes:
1. We have to move the initialization before the lightningd main trace_start,
   since that uses pseudorand().
2. To make the results stable, we need to use per-caller values to randbytes().
   Otherwise external timing changes the call order.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
global: replace randombytes_buf() with randbytes() wrapper.

This allows us to override it for deterministic results.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ShahanaFarooqui reviewed (commented):
@rustyrussell The changes in `./tests/autogenerate-rpc-examples.py` look good at first glance, but I am unable to test them locally due to missing test data files:
```
./tests/data/autogenerate-bitcoin-blocks.json
./tests/data/autogenerate-bitcoind-wallet.dat
```

Could you please add these files or provide instructions on how to generate them for local testing?

rustyrussell committed:
autogenerate-rpc-examples.py: disable entropy for generation.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: remove example mangling.

We temporarily add a flake8 exception so it doesn't complain about all
the now-unused variables.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: remove unused vars

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
tests/autogenerate-rpc-examples.py: always use largest UTXO for fundchannel example.

Now order has changed, the first one is insufficient; this is more robust.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: canned blocks.

Unless REGENERATE_BLOCKCHAIN is true, in which case we make them.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: ensure blockheights are consistent.

This makes our transaction generation (nlocktime) consistent, as well as our htlcs
for payment.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: use fixed port numbers.

Move the elements config writing into an ElementsD.set_port() override. This makes many of our information outputs consistent.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: more refinement.

1. Don't mangle bookkeeper events.
2. Use l3, not l1, for listtransactions: it's more stable.
3. Use l2, not l1, for listpeerchannels and listchannels.
3. Use l2, not l2, for listfunds.

(l1 has a lot of variation).

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: more block generation and synchronization.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
autogenerate-rpc-examples.py: rewrite problematic examples.

And hand in CLN_NEXT_VERSION to replace the version string.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
CI: re-enable docs examples checks.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ShahanaFarooqui commented:
Now `uv run make update-doc-examples` is throwing error:

```
15:12:18 - ERROR - Error in setting up nodes: txid f160d78f39a5ccb96c1439333787b98f1a37c1bb3765b34c74047c8895f4e397 not found
15:12:18 - ERROR - txid f160d78f39a5ccb96c1439333787b98f1a37c1bb3765b34c74047c8895f4e397 not found
Traceback (most recent call last):
  File "/home/shahana/workspace/lightning/tests/autogenerate-rpc-examples.py", line 1443, in test_generate_examples
    l1, l2, l3, l4, l5, l6, c12, c23, c25 = setup_test_nodes(node_factory, bitcoind, regenerate_blockchain)
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shahana/workspace/lightning/tests/autogenerate-rpc-examples.py", line 331, in setup_test_nodes
    c12, c12res = l1.fundchannel(l2, FUND_CHANNEL_AMOUNT_SAT)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shahana/workspace/lightning/contrib/pyln-testing/pyln/testing/utils.py", line 1194, in fundchannel
    assert False, f"txid {res['txid']} not found"
           ^^^^^
AssertionError: txid f160d78f39a5ccb96c1439333787b98f1a37c1bb3765b34c74047c8895f4e397 not found
```

And `uv run make update-doc-examples-newchain` is throwing:

```
15:10:44 - ERROR - Error in generating transactions examples: RPC call failed: method: fetchbip353, payload: {'address': 'send.some@satsto.me'}, error: {'code': -32700, 'data': None, 'message': 'failed to fetch payment instructions: HrnResolutionError("Multiple TXT records existed for the HRN, which is invalid")'}
15:10:44 - ERROR - RPC call failed: method: fetchbip353, payload: {'address': 'send.some@satsto.me'}, error: {'code': -32700, 'data': None, 'message': 'failed to fetch payment instructions: HrnResolutionError("Multiple TXT records existed for the HRN, which is invalid")'}
Traceback (most recent call last):
  File "/home/shahana/workspace/lightning/tests/autogenerate-rpc-examples.py", line 1444, in test_generate_examples
    c23_2, c23res2, c34_2, inv_l11, inv_l21, inv_l22, inv_l31, inv_l32, inv_l34 = generate_transactions_examples(l1, l2, l3, l4, l5, c25, bitcoind)
                                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shahana/workspace/lightning/tests/autogenerate-rpc-examples.py", line 463, in generate_transactions_examples
    update_example(node=l1, method='fetchbip353', params={'address': 'send.some@satsto.me'}, description=['Example of fetching BIP-353 payment details.'])
  File "/home/shahana/workspace/lightning/tests/autogenerate-rpc-examples.py", line 260, in update_example
    response = node.rpc.call(method, params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shahana/workspace/lightning/contrib/pyln-testing/pyln/testing/utils.py", line 811, in call
    res = LightningRpc.call(self, method, payload, cmdprefix, filter)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shahana/workspace/lightning/contrib/pyln-client/pyln/client/lightning.py", line 428, in call
    raise RpcError(method, payload, resp['error'])
pyln.client.lightning.RpcError: RPC call failed: method: fetchbip353, payload: {'address': 'send.some@satsto.me'}, error: {'code': -32700, 'data': None, 'message': 'failed to fetch payment instructions: HrnResolutionError("Multiple TXT records existed for the HRN, which is invalid")'}
```

ShahanaFarooqui committed:
make: Fix repeat-dpc-examples  script

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: parameterized base port and cln next version

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: added 17 new RPC to ignore list

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: added wait functions like wait_for_htlcs_settled, wait_for_wallet_txs

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: added feature to be able to rewrite the request in Rewriter

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: fixed recover examples via `lightning-hsmtool getsecret`

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: added more rewrite for other RPCs

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: added sorting for some list example to stabilise the response

And flake fixes

ShahanaFarooqui committed:
autogenerate-rpc-examples.py: pin the full sorted UTXO set for multiwithdraw examples

ShahanaFarooqui committed:
doc: generate all examples with newchain

ShahanaFarooqui committed:
make: Pinning PATH in doc scripts so generation always uses the tree's own binaries

ShahanaFarooqui committed:
common: seed deterministic RNG from basename of argv0, not full path

dev_override_randbytes() hashes argv0 into the CLN_DEV_ENTROPY_SEED
stream, so every plugin and subdaemon gets a distinct seed.  But
lightningd execs them with absolute paths, so the "deterministic"
stream silently depended on where the source tree was checked out:
canned blocks generated in one directory would not replay in another
(funding tx output order and nlocktime fuzz diverge, so txids change).
This is why check-doc-examples passed locally but failed in CI.

Hash only the basename: binaries still get distinct seeds, but the
stream no longer depends on the checkout path.

ShahanaFarooqui committed:
common: don't consume deterministic randbytes stream for trace ids

Trace/span ids now come from a plain counter when CLN_DEV_ENTROPY_SEED is active, so tracing (which CI's  HAVE_USDT=1 build enables by default) no longer shifts the deterministic RNG stream that PSBT serial ids and locktime fuzz are drawn from.

ShahanaFarooqui committed:
autogenerate-rpc-examples: sanitize plugin paths in plugin examples

ShahanaFarooqui committed:
common: don't trigger siphash_seed() init from span hashing either.

Previous commit stopped trace ids consuming the deterministic
randbytes stream, but tracing still touched it once: span_key_hash()
calls siphash_seed(), whose lazy init draws 16 bytes from the stream.
Since randbytes call sites are assigned their offset blocks in
first-use order, that single draw shifts every later draw in the
process by 1000 when tracing is active (HAVE_USDT builds, as in CI):
keysend preimages and invoicerequest payer keys then differ from a
build with tracing disabled, showing up as an 11-file doc-examples
diff.

Use a cheap multiplicative hash for span keys when the deterministic
override is active, so tracing touches the RNG stream zero times.

ShahanaFarooqui committed:
CI: Add debug and upload artifacts on failure

ShahanaFarooqui committed:
tests: use bitcoind.conf_file instead of hardcoded bitcoin.conf in RawProxy mocks

These tests built their RawProxy against a hardcoded bitcoin.conf in the node's datadir. That only worked on liquid-regtest by accident: BitcoinD.__init__ used to write bitcoin.conf even when the node was actually an ElementsD. Now that config writing is deferred to set_port() and ElementsD only writes elements.conf, the mocks found no config and no .cookie, so their RPC calls failed and bcli's getblock timed out, leaving lightningd stuck during restart.

Use bitcoind.conf_file, which points at the config the running daemon actually uses (elements.conf on liquid-regtest, bitcoin.conf on egtest).

rustyrussell reviewed (commented):
Ack 00dc5214c3d1d36bc5054a809af2b796b157e7cb

ShahanaFarooqui reviewed (approved):
Ack 00dc5214c3d1d36bc5054a809af2b796b157e7cb

madelinevibes reviewed (approved):
ack 00dc5214c3d1d36bc5054a809af2b796b157e7cb

nGoline commented on .github/workflows/ci.yaml:
```diff
@@ -839,11 +838,57 @@ jobs:
           find . -type f -print0 | xargs -0 touch -d yesterday
           tar xaf cln-${{ matrix.CFG }}.tar.gz
 
+      - name: Debug check-doc-examples inputs
+        run: |
+          echo "== lightning-hsmtool =="
+          which lightning-hsmtool || echo "NOT ON PATH"
+          find . -name "lightning-hsmtool*" -not -path "./.git/*" 2>/dev/null
+
+          echo "== autogenerate-bitcoind-wallet.dat =="
+          ls -la tests/data/autogenerate-bitcoind-wallet.dat 2>&1
+
+          echo "== autogenerate-bitcoin-blocks.json =="
+          ls -la tests/data/autogenerate-bitcoin-blocks.json 2>&1
+          echo "-- number of blocks --"
+          python3 -c "import json; print(len(json.load(open('tests/data/autogenerate-bitcoin-blocks.json'))))" 2>&1
+          echo "-- first 300 chars --"
+          head -c 300 tests/data/autogenerate-bitcoin-blocks.json
+          echo ""
+
+          echo "== environment fingerprint =="
+          # Determinism discriminators: HAVE_USDT toggles tracing, which
+          # must not perturb the deterministic RNG stream.
+          grep -E 'HAVE_USDT|HAVE_SQLITE3' config.vars || true
+          bitcoind --version | head -1 || true
+          dpkg -l systemtap-sdt-dev 2>/dev/null | tail -1 || true
+
```
Don't forget to remove this debug scaffolding.

nGoline commented on contrib/pyln-testing/pyln/testing/utils.py:
```diff
@@ -935,6 +944,7 @@ def __init__(self, node_id, lightning_dir, bitcoind, executor, may_fail=False,
                  executable=None,
                  bad_notifications=False,
                  old_hsmsecret=None,
+                 no_entropy=False,
```
The naming seems backwards. `no_entropy=True` actually enables the deterministic entropy seed (`CLN_DEV_ENTROPY_SEED`), so the name reads as the opposite of what it does. Consider `deterministic_entropy=` or `entropy_seed=`, or even add a docstring line here clarifying "no real entropy -> use a fixed dev seed."

nGoline commented on contrib/pyln-testing/pyln/testing/utils.py:
```diff
@@ -1000,6 +1011,8 @@ def __init__(self, node_id, lightning_dir, bitcoind, executor, may_fail=False,
         # Avoid test flakes cause by this option unless explicitly set.
         if self.cln_version >= "v24.11":
             self.daemon.opts.update({"autoconnect-seeker-peers": 0})
+        if no_entropy:
+            self.daemon.env["CLN_DEV_ENTROPY_SEED"] = str(node_id)
```
Looking at the behaviour makes it clear the variable does the opposite.

nGoline commented on contrib/pyln-testing/pyln/testing/utils.py:
```diff
@@ -1827,10 +1842,14 @@ def get_node(self, node_id=None, options=None, dbfile=None,
         node_id = self.get_node_id() if not node_id else node_id
-        port = reserve_unused_port()
-        grpc_port = self.get_unused_port() if unused_grpc_port else None
+        if base_port:
+            port = base_port + node_id * 2 - 1
+            grpc_port = base_port + node_id * 2
```
With `base_port set`, `port/grpc_port` are computed deterministically and intentionally not added to `reserved_ports` (line 1876). That's fine for a single serial doc-gen run, but a collision here surfaces as a confusing bind error instead of a clean reservation retry. Worth a one-line comment stating the assumption (single serial run over a fixed range).

nGoline commented on contrib/pyln-testing/pyln/testing/utils.py:
```diff
@@ -1854,7 +1873,8 @@ def get_node(self, node_id=None, options=None, dbfile=None,
 
         self.nodes.append(node)
-        self.reserved_ports.append(port)
+        if not base_port:
+            self.reserved_ports.append(port)
```
`base_port` ports aren't reserved or collision-checked.

nGoline commented on common/randbytes.c:
Pre-existing, but flagging since determinism is the whole point of this PR: `randbytes_overridden()` keys off `dev_seed != 0`, and the basename change alters the `siphash24` input. If the hash ever lands on exactly `0`, the override silently disables and output becomes non-deterministic rather than erroring. A dedicated `dev_overridden` bool would be more robust than sentinel-zero.

nGoline commented on common/randbytes.c:
```diff
@@ -48,15 +49,23 @@ void randbytes_(void *bytes, size_t num_bytes, u64 *offset)
+	const char *base;
 	assert(!used);
 
+	/* Hash only the basename: binaries still get distinct seeds, but
+	* the stream no longer depends on the checkout path. Plugins and
+	* subdaemons are exec'd with absolute paths: hash only the basename,
+	* so the stream doesn't depend on where the source tree lives */
```
nit: the comment says "hash only the basename" twice (lines 59 and 61–62 restate it), and lines 60/62 are indented one space short of the block (`\t*` vs `\t *`).
